modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/CursorSubscription.java

```diff
@@ -0,0 +1,185 @@
+    private int responseIndex;
+
+    /** Amount of entries requested by the subscriber. */
+    private long demand;
+
+    CursorSubscription(
+            RaftGroupService raftService,
+            MetaStorageCommandsFactory commandsFactory,
+            ExecutorService executorService,
+            IgniteUuid cursorId,
+            Subscriber<? super Entry> subscriber
+    ) {
+        this.raftService = raftService;
+        this.commandsFactory = commandsFactory;
+        this.executorService = executorService;
+        this.cursorId = cursorId;
+        this.subscriber = subscriber;
+    }
+
+    @Override
+    public void request(long n) {
+        if (n <= 0) {
+            onError(new IllegalArgumentException("Requested amount must be greater than zero, got: " + n));
+
+            return;
+        }
+
+        if (isDone) {
+            return;
+        }
+
+        // Increase the number of requested items.
+        demand += n;
+
+        if (demand <= 0) {
+            onError(new IllegalArgumentException("Long overflow"));
+
+            return;
+        }
+
+        // Start the processing if it has not been started yet (async operation).
+        if (cachedResponse == null) {
+            requestNextBatch();
+        }
+    }
+
+    /**
+     * Does the actual request processing. Must always be called on {@link CursorSubscription#executorService}.
+     */
+    private void processRequest() {
+        if (isDone) {
+            return;
+        }
+
+        try {
+            assert cachedResponse != null;
+
+            List<Entry> entries = cachedResponse.entries();
+
+            while (demand > 0 && !isDone) {
+                if (responseIndex < entries.size()) {
+                    subscriber.onNext(entries.get(responseIndex));
+
+                    responseIndex++;
+                    demand--;
+                } else {
+                    if (cachedResponse.hasNextBatch()) {
+                        requestNextBatch();
+                    } else {
+                        isDone = true;
+
+                        subscriber.onComplete();
+                    }
+
+                    return;
+                }
+            }
+        } catch (Exception e) {
+            onError(e);
+        }
+    }
+
+    private void requestNextBatch() {
+        NextBatchCommand command = commandsFactory.nextBatchCommand()
+                .cursorId(cursorId)
+                .batchSize(BATCH_SIZE)
+                .build();
+
+        raftService.<BatchResponse>run(command)
+                .whenCompleteAsync((resp, e) -> {
+                    if (e == null) {
+                        cachedResponse = resp;
+                        responseIndex = 0;
+
+                        processRequest();
+                    } else {
+                        onError(e);
+                    }
+                }, executorService);
+    }
+
+    @Override
+    public void cancel() {
+        isDone = true;
```

Comment:
And what about this one? Seems that it's possible and valid to cancel subscriptions within some other threads, e.g. components.stop one.

Reply:
Can you explain how that could happen? I also expect this method to be called from the same thread (e.g. during `onNext` or `onError`).

Reply:
I mean following (pseudocode)
```
    private volatile Subscription msRangeSub;
    
        ms.range().subscribe(new Subscriber<Object>() {

        @Override
        public void onSubscribe (Subscription subscription){
            this.msRangeSub = subscription;
        }
    }
    
    stop() {
        msRangeSub.cancel();
    }
```

Reply:
Shame on you, you are right.
`Subscription.request and Subscription.cancel MUST only be called inside of its Subscriber context.`
https://github.com/reactive-streams/reactive-streams-jvm/